cluster/src/main/scala/org/apache/spark/memory/SnappyUnifiedMemoryManager.scala

```diff
@@ -349,17 +356,37 @@ class SnappyUnifiedMemoryManager private[memory](
       capacity, changeOwner)
   }
 
-  def tryExplicitGC(numBytes: Long): Unit = {
+  private def tryExplicitGC(numBytes: Long): Unit = {
```

Comment:
Explicit GC was originally added for off-heap case where reference count of some DirectByteBuffers did not go down to zero despite the fact that no one is holding references to them. Initially there were some code issues that caused that in some situations but as of now the only known cases are exceptions where explicit release() of the DirectByteBuffer (which in turn calls Unsafe.freeMemory) has not been added in a finally block all over the code.

The original design of the "new" off-heap was such that there were no leaks as such even if the explicit reference count does not go down to zero in all possible paths. That dependence had caused tons of code changes with the GemFire off-heap implementation and issues where dangling reference count remains (even now we get occasional failures in off-heap unit tests). The new implementation uses the PhantomReference route where GC will queue up the reference object once the underlying ByteBuffer object is collected. This is the way Java's ByteBuffer.allocateDirect operates where there is no release() API to being with and it depends on the garbage collector to detect that ByteBuffer is garbage and queues the PhantomReference where the queue cleaner then does the freeMemory internally. See FreeBufferBase.run in ManagedDirectBufferAllocator.java which overrides JDK's DirectByteBuffer Cleaner by reflection to do UMM accounting when buffer is freed (which is the reason off-heap memory accounting is always accurate with no danger of missed accounting in some paths). The one down-side of this is that the heap pressure of DirectByteBuffer object is small (~200 bytes) so they themselves piling up due to "leak" will usually not cause GC to kick in. So if there have been tons of exception cases and there is no one else to pressure the heap, then even explicit GC could get ignored and one could still see unexpected LMEs (though as of now there is no such case seen in practise).

So the explicit release() usage in our implementation is really only a convenience though in practise it helps eliminate stray LME's due to GC still not having kicked in. The purpose of the explicit GC call is to have one more go at it before throwing LME because we almost never expect off-heap to fail with OOME since it will overflow as much as necessary (except for the case where total size cannot even fit the buffers needed by threads during inserts i.e. 24MB * total-no-of-spark-cores). With the new SHA implementation there is another case where it can cause genuine LME since those cannot overflow to disk yet.

Reply:
ok. Thanks for the detailed insight.